bindings/azure/signalr/signalr.go

```diff
@@ -0,0 +1,195 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package signalr
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/gbrlsnchs/jwt/v3"
+	"github.com/pkg/errors"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	errorPrefix         = "azure signalr error:"
+	logPrefix           = "azure signalr:"
+	connectionStringKey = "connectionString"
+	hubKey              = "hub"
+	groupKey            = "group"
+	userKey             = "user"
+)
+
+// NewSignalR creates a new pub/sub based on Azure SignalR
+func NewSignalR() *SignalR {
+	return &SignalR{
+		tokens:     make(map[string]signalrCachedToken),
+		httpClient: &http.Client{Timeout: 30 * time.Second},
+	}
+}
+
+type signalrCachedToken struct {
+	token      string
+	expiration time.Time
+}
+
+// SignalR is an output binding for Azure SignalR
+type SignalR struct {
+	endpoint   string
+	accessKey  string
+	version    string
+	hub        string
+	tokens     map[string]signalrCachedToken
+	httpClient *http.Client
+}
+
+// Init is responsible for initializing the SignalR output based on the metadata
+func (s *SignalR) Init(metadata bindings.Metadata) error {
+	connectionString, ok := metadata.Properties[connectionStringKey]
+	if !ok || connectionString == "" {
+		return fmt.Errorf("missing connection string")
+	}
+
+	if hub, ok := metadata.Properties[hubKey]; ok && hub != "" {
+		s.hub = hub
+	}
+
+	// Expected: Endpoint=https://<servicename>.service.signalr.net;AccessKey=<access key>;Version=1.0;
+	connectionValues := strings.Split(strings.TrimSpace(connectionString), ";")
+	for _, connectionValue := range connectionValues {
+		if i := strings.Index(connectionValue, "="); i != -1 {
+			k := connectionValue[0:i]
+			switch k {
+			case "Endpoint":
+				s.endpoint = connectionValue[i+1:]
```

Comment:
Is it worth checking i+1 exists before indexing into the array? i.e. `Endpoint=;`

Reply:
Good catch, thanks @jjcollinge